cmd/kinput2.man
.\" $Id: kinput2.man,v 1.20 1999/09/13 08:40:31 ishisone Exp $
.TH KINPUT2 1 "24 March 1999" "X Version 11"
.SH NAME
kinput2 \- kanji input server for X11
.SH SYNOPSIS
\fBkinput2\fP [-options...]
.SH DESCRIPTION
.I Kinput2
is an X window application to input Japanese text.
It acts as a front-end for other applications that want
kana-kanji conversion.
.PP
When invoked, after some initialization (which will take about a few to
30 seconds depending on your machine) \fIkinput2\fP waits quietly for a
Japanese text input request from another client (i.e. no windows
appear).
When \fIkinput2\fP receives a request, it pops up a window and
starts conversion process. It sends the converted text back to the
client when the text is fixed.
.PP
.I Kinput2
has some big improvements over its predecessor, kinput, which
was contributed to X11R4.
.TP 8
\fBOver-the-spot/On-the-spot Input\fP
These features enable conversion process to be done at cursor position,
avoiding unnecessary movement of eyes between cursor position and conversion
window.
.I Kinput2
also supports root window style input and off-the-spot style input.
.TP 8
\fBMultiple Protocol Support\fP
.I Kinput2
supports several conversion protocols between front-end and
client. Supported protocols are:
.nf
	kinput protocol
	kinput2 protocol
	Matsushita jinput protocol
	Sony xlc protocol
	XIMP protocol
	X Input Method Protocol (X Consortium standard)
.fi
.TP 8
\fBMultiple Conversion Engine Support\fP
.I Kinput2
can use 4 different conversion engines, Wnn, Canna, Sj3 and Atok.
You can choose one at compilation time, or at run time.
.TP 8
\fBKinput Compatibility\fP
.I Kinput2
is fully upward-compatible with kinput, so applications which use
kinput as the front-end can also use \fIkinput2\fP.
In this case, the applications cannot take advantage of over-the-spot
conversion, though.
.SH OPTIONS
.I kinput2
accepts all of the standard X Toolkit command line options.
The following options are accepted as well.
.TP 8
\fB\-wnn\fP
This option specifies that Wnn jserver is used as the conversion engine.
.TP 8
\fB\-jserver\fP
This option specifies the hostname of the jserver to be used for conversion.
.TP 8
\fB\-ccdef\fP
This option specifies the input character conversion rule file.
.TP 8
\fB\-wnnenvrc4\fP
This option specifies the Wnn convertion environment initialization
file (wnnenvrc) for Wnn version 4.
.TP 8
\fB\-wnnenvrc6\fP
This option is similar to \-wnenvrc4, but for Wnn version 6.
.TP 8
\fB\-canna\fP
This option specifies that Canna conversion server (irohaserver) is used as
the conversion engine.
.TP 8
\fB\-cannaserver\fP
This option specifies the hostname of Canna conversion server.
.TP 8
\fB\-cannafile\fP
This option specifies Canna conversion customization file.
.TP 8
\fB\-sj3\fP
This option specifies that Sj3 conversion server (sj3serv) is used as
the conversion engine.
.TP 8
\fB\-sj3serv\fP
This option specifies the hostname of the primary sj3 conversion server.
.TP 8
\fB\-sj3serv2\fP
This option specifies the hostname of the secondary sj3 server. The secondary
server is used when \fIkinput2\fP cannot connect to the primary server.
.TP 8
\fB\-atok\fP
This option specifies that Atok server is used as the conversion engine.
.TP 8
\fB\-atokserver\fP
This option specifies the hostname of the atok server to be used for conversion.
.TP 8
\fB\-font\fP \fIascii-font\fP
This option specifies the default font to be used for displaying ASCII text.
.TP 8
\fB\-kanjifont\fP \fIkanji-font\fP
This option specifies the default font to be used for displaying Kanji text.
.TP 8
\fB\-kanafont\fP \fIkana-font\fP
This option specifies the default font to be used for displaying Kana text.
.TP 8
\fB\-/+kinput\fP
This option enables/disables use of kinput protocol family (i.e.
kinput protocol, kinput2 protocol, jinput protocol and xlc protocol).
.TP 8
\fB\-/+ximp\fP
This option enables/disables use of XIMP protocol.
.TP 8
\fB\-/+xim\fP
This option enables/disables use of X Input Method Protocol.
.TP 8
\fB\-bc\fP
Kinput protocol, which is one of the communication protocols
\fIkinput2\fP supports,
uses X window's selection mechanism to notify other clients of
front-end's existence.
This option forces \fIkinput2\fP to use selection ``JAPANESE_CONVERSION''
(which is not a standard selection name) as well as ``_JAPANESE_CONVERSION''.
This is for backward compatibility, since many of the applications that
connect with kinput still use this non-standard selection.
.TP 8
\fB\-tbheight\P \fItitle-bar-height\fP
This option specifies the height (in pixels)  of the titlebar
attached to popup shells such as candidate selection window.
.I Kinput2
uses this value to compute the correct popup location of these windows.
.SH WIDGET TREE
Below is the widget hierarchy of \fIkinput2\fP.
.nf
.ta 0.5i 1.0i 1.5i 2.0i 2.5i 3.0i 3.5i 4.0i
	Kinput2  kinput2
		ConversionManager  convmanager
			KinputProtocol  kinputprotocol
			XimpProtocol  ximpprotocol
			IMProtocol  improtocol
			SeparateConversion  converter
				CcWnn  inputObj
				Canna  inputObj
				Sj3  inputObj
				JpWcharDisplay  displayObj
				Form  form
					ICLabel  mode
						JpWcharDisplay  display
					Canvas  text
				TransientShell  selectionShell
					CandidatePanel  selection
						JpWcharDisplay  display
				TransientShell  auxShell
					AuxPanel  aux
						JpWcharDisplay  display
			OffTheSpotConversion  converter
				CcWnn  inputObj
				Canna  inputObj
				Sj3  inputObj
				JpWcharDisplay  displayObj
				AdoptedShell  modeShell
					ICLabel  mode
						JpWcharDisplay display
				CanvasShell  text
				TransientShell  selectionShell
					CandidatePanel  selection
						JpWcharDisplay  display
				TransientShell  auxShell
					AuxPanel  aux
						JpWcharDisplay  display
			OverTheSpotConversion  converter
				CcWnn  inputObj
				Canna  inputObj
				Sj3  inputObj
				JpWcharDisplay  displayObj
				TransientShell  selectionShell
					CandidatePanel  selection
						JpWcharDisplay  display
				TransientShell  auxShell
					AuxPanel  aux
						JpWcharDisplay  display
				AdoptedShell  modeShell
					ICLabel  mode
						JpWcharDisplay  display
				TransientShell  modeShell
					ICLabel  mode
						JpWcharDisplay  display
				CanvasShell  text
.fi
.SH RESOURCES
There are too many resources to describe here, but you don't have to
know most of them.
So here is a brief description of resources which you might want to set.
.PP
\fBKinput2\fP application resource:
.RS 4
.TP 4
\fBconversionEngine\fP
This resource specifies which conversion engine to use.
You can specify one of ``wnn'', ``canna'', ``sj3'' or ``atok''.
.RE
.PP
.B ConversionControl
widget (which is a superclass
of SeparateConversion, OffTheSpotConversion and OverTheSpotConversion)
resource:
.RS 4
.TP 4
\fBsendbackKeyPress\fP
If this resource is ``true'', the KeyPress events intercepted but unused
by \fIkinput2\fP are sent back to the client when there is no conversion text.
Although this feature is convenient, it might confuse some clients
since none of corresponding KeyRelease events are sent.
Also, since the ``send_event'' flag of those events are true, some
clients do not accept events sent back by \fIkinput2\fP.
For example, kterm doesn't accept those events without specifying
.nf
	KTerm*allowSendEvents: true
.fi
in a resource file.
.br
The default of \fBsendbackKeyPress\fP resource is \fBtrue\fP.
.TP 4
\fBtitlebarHeight\fP
This resource specifies the height of a title bar which is attached
to various popup shell windows, such as candicate selection.
When computing the location of these windows,
.I kinput2
uses this value to pop up them at correct position.
.br
The default is \fB0\P.
.RE
.PP
.B OverTheSpotConversion
widget resources:
.RS 4
.TP 4
\fBshrinkWindow\fP
If this resource is ``true'', the width of on-the-spot conversion window
shrinks according to the length of the text on it. Otherwise,
the window never shrinks.
If the performance of Xserver on window resizing is poor,
you may consider setting this resources to false.
The default is \fBtrue\fP.
.TP 4
\fBmodeLocation\fP
This resource specifies the location of status widget in case of 
over-the-spot type conversion.
Unless client specifies the location explicitly, the location of
the status widget is determined by this resource.
If the value of this resource is ``topleft'', the widget is placed at the
top left of the client window. You can also specify 
``topright'', ``bottomleft'' and ``bottomright''.
If the value is ``tracktext'', the widget is placed just under the cursor
position, and moves according to the cursor. However, if the
region available for conversion is too small, status is placed at
the bottom left of the client area.
If the value is ``none'', the status widget will never appear.
.br
The default is ``tracktext''.
.TP 4
\fBmodeBorderForeground\fP
If this resource is ``true'', the color of status widget is forced
to be same as the foreground color of the widget.
.br
The default is \fBfalse\fP.
.TP 4
\fBuseOverrideShellForMode\fP
If this resource is ``true'', the status widget becomes an instance of
OverriddeShell, instead of the default TransientShell.
Turn this resource on if you encounter a strange phenomenon that
the status widget keeps flickering and you cannnot enter any text.
This is caused by (incorrect) focus management of your window manager,
and setting this resource will stop any window management for the widget.
The result is that the widget will be totally ignored by the window manager,
meaning you cannot move, resize, raise nor lower the widget.
.br
The default is \fBfalse\fP.
.RE
.PP
.B OnTheSpotConversion
widget resources:
.RS 4
.TP 4
\fBfeedbackAttributes\fP
This resource specifies the display attributes of preedit string when
on-the-spot input style is used. In on-the-spot style, it is a client's
responsibility to draw preedit string, and kinput2 sends some hints
on how it should be drawn, which is controled by this resource.

NEWS
Major changes from version 3.0 to 3.1:

* Kinput2 now supports ATOK.  The source code is kindly contributed
  by Justsystem Corporation.  Thanks!
  To compile for Atok, define symbol `UseAtok' in Kinput2.conf.

* Minimum support of Wnn7 is implemented.

* Two IC attributes, XNPreeditState and XNResetState both of which
  are introduced by X11R6, are supported.

* Kinput2's `egg'-emulation is improved.

* A workaround for bugs in some Xlib implementations has been added.

* Previously, when kinput2 cannot connect to jserver, kinput2 seems
  to be frozen (actually it keeps retrying).  Now kinput2 beeps and
  ends conversion in such a case.  The next time you start conversion,
  it tries to connect again.


Major changes from version 2.0 fix4 (the last official release) to
version 3.0:

* Kinput2 now supports Wnn6.  The source code is kindly contributed by
  MORIBE, Hideyuki.  Thanks!  To compile for Wnn6, define symbol
  `UseWnn6' in Kinput2.conf.

  For those who use both Wnn4 and Wnn6, distinct wnn initialization
  files (wnnenvrc) for Wnn4 jserver and Wnn6 jserver can be specified.
  Kinput2 checks the server version at run time, and selects an
  appropriate one.  Use `*.wnnEnvrc4' resource for specifying
  initialization file for Wnn4, `*.wnnEnvrc6' for Wnn6.

  Due to Wnn6 support, Wnn4 implementation part saw major changes.
  We've tried to keep user visible changes small, though.

* On-the-spot input style is available for Ximp and XIM protocols.
  The code is contributed by Kazuki YASUMATSU.  Thank you!
  For XIM protocol, following styles are now supported:
  	XIMPreeditPosition | XIMStatusArea (over-the-spot style)
  	XIMPreeditPosition | XIMStatusNothing (over-the-spot style)
  	XIMPreeditArea | XIMStatusArea (off-the-spot style)
	XIMPreeditCallbacks | XIMStatusCallbacks (on-the-spot style)
	XIMPreeditCallbacks | XIMStatusNothing (on-the-spot style)
	XIMPreeditNothing | XIMStatusNothing (root-window style)

* Kinput2 now supports an IC attribute called FilterEvents.  Some
  applications seem to require the input server to support this
  attribute, or they go crazy if not.

* Finally, word registration facility is available for Wnn.  A new
  fuction called `register' is added, and bound to F5 and ^T by
  default. (If you're using ccdef.kinput2.egg, that's F4) This
  function is supposed to invoke word registration utility called
  `tkwnnreg'.  You need a separate package for the utility, currently
  available from:
	ftp://ftp.sra.co.jp/pub/lang/tcl/contrib/tcl-wnn-X.Y.tar.gz
  X.Y is 0.3 as of this writing.  It is still alpha-quality, but I'll
  work on it once I finish release work of kinput2.

* Kinput2's `egg'-emulation has been improved.

* You can specify preedit font and status font independently when XIM
  protocol is used.  Previously preedit font was used for both preedit
  area and status area.

* Now kinput2's mode shell (the small status window appeared when
  over-the-spot input is active) has Mwm hints that says "I need no
  titlebar".

* A new boolean resource `useOverrideShellForMode' is provided to use
  overrideShell for the mode displaying widget, instead of the default
  transientShell.  This is needed for some window managers which
  ignores kinput2's input focus preference, causing flickering of mode
  display.




cmd/kinput2.c
/*
 *	kinput2
 */

/*
 * Copyright (C) 1991 by Software Research Associates, Inc.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Software Research Associates not be
 * used in advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.  Software Research
 * Associates makes no representations about the suitability of this software
 * for any purpose.  It is provided "as is" without express or implied
 * warranty.
 *
 * Author:  Makoto Ishisone, Software Research Associates, Inc., Japan
 *		ishisone@sra.co.jp
 */

#ifndef lint
static char	*rcsid = "$Id: kinput2.c,v 1.46 2002/01/06 15:13:38 ishisone Exp $";
#endif

#include <stdio.h>
#include <signal.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include "patchlevel.h"
#include "AsyncErr.h"
#include "MyDispatch.h"
#include "IOECall.h"
#include "ConvMgr.h"
#include "KIProto.h"
#include "XimpProto.h"
#include "IMProto.h"

#if !defined(USE_WNN) && !defined(USE_CANNA) && !defined(USE_SJ3) && !defined(USE_ATOK)
#define USE_WNN			/* default */
#endif

#ifdef USE_WNN
#include "CcWnn.h"
#include "WcharDisp.h"
#endif
#ifdef USE_CANNA
#include "Canna.h"
#include "WcharDisp.h"
#endif
#ifdef USE_SJ3
#include "Sj3.h"
#include "WcharDisp.h"
#endif
#ifdef USE_ATOK
#include "Atok.h"
#include "WcharDisp.h"
#endif

#include "DebugPrint.h"

int	debug_all;

/* application resource */
typedef struct {
    char *conversionEngine;
    int debugLevel;
    Boolean useKinputProtocol;
    Boolean useXimpProtocol;
    Boolean useXIMProtocol;
    Boolean appdefs_loaded;
} AppResRec, *AppResP;

static AppResRec appres;

static XtResource app_resources[] = {
    { "conversionEngine", "ConversionEngine", XtRString, sizeof(String),
      XtOffset(AppResP, conversionEngine), XtRString, (XtPointer)"" },
    { "debugLevel", "DebugLevel", XtRInt, sizeof(int),
      XtOffset(AppResP, debugLevel), XtRImmediate, (XtPointer)0 },
    { "useKinputProtocol", "UseKinputProtocol", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, useKinputProtocol), XtRImmediate, (XtPointer)True },
    { "useXimpProtocol", "UseXimpProtocol", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, useXimpProtocol), XtRImmediate, (XtPointer)True },
    { "useXIMProtocol", "UseXIMProtocol", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, useXIMProtocol), XtRImmediate, (XtPointer)True },
    { "ki2AppDefsLoaded", "Ki2AppDefsLoaded", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, appdefs_loaded), XtRImmediate, (XtPointer)True },
};

static String fallback_resources[] = {
    "*ki2AppDefsLoaded: false",
    "Kinput2.mappedWhenManaged: false",
    "Kinput2.width: 1",
    "Kinput2.height: 1",
    NULL,
};

static XrmOptionDescRec	options[] = {
    {"-bc",		"*KinputProtocol.backwardCompatible", XrmoptionNoArg,	"True"},
    {"-font",		"*JpWcharDisplay.font",	XrmoptionSepArg,	NULL},
    {"-kanjifont",	"*JpWcharDisplay.kanjiFont", XrmoptionSepArg,	NULL},
    {"-kanafont",	"*JpWcharDisplay.kanaFont", XrmoptionSepArg,	NULL},
    {"-kinput",		".useKinputProtocol",	XrmoptionNoArg,		"on"},
    {"+kinput",		".useKinputProtocol",	XrmoptionNoArg,		"off"},
    {"-ximp",		".useXimpProtocol",	XrmoptionNoArg,		"on"},
    {"+ximp",		".useXimpProtocol",	XrmoptionNoArg,		"off"},
    {"-xim",		".useXIMProtocol",	XrmoptionNoArg,		"on"},
    {"+xim",		".useXIMProtocol",	XrmoptionNoArg,		"off"},
    {"-tbheight",	"*ConversionControl.titlebarHeight",
						XrmoptionSepArg,	NULL},
#ifdef DEBUG
    {"-debug",		".debugLevel",		XrmoptionNoArg,		"1"},
    {"-trace",		".debugLevel",		XrmoptionNoArg,		"10"},
    {"-debuglevel",	".debugLevel",		XrmoptionSepArg,	NULL},
#endif
#ifdef USE_WNN
    {"-wnn",		".conversionEngine",	XrmoptionNoArg,		"wnn"},
    {"-jserver",	"*CcWnn.jserver",	XrmoptionSepArg,	NULL},
    {"-ccdef",		"*CcWnn.ccdef",		XrmoptionSepArg,	NULL},
    {"-wnnenvname",	"*CcWnn.wnnEnvname",	XrmoptionSepArg,	NULL},
    {"-wnnenvrc4",	"*CcWnn.wnnEnvrc4",	XrmoptionSepArg,	NULL},
    {"-wnnenvrc6",	"*CcWnn.wnnEnvrc6",	XrmoptionSepArg,	NULL},
    {"-wnnenvrc",	"*CcWnn.wnnEnvrc",	XrmoptionSepArg,	NULL},
#endif
#ifdef USE_CANNA
    {"-canna",		".conversionEngine",	XrmoptionNoArg,		"canna"},
    {"-cannaserver",	"*Canna.cannahost",	XrmoptionSepArg,	NULL},
    {"-cs",		"*Canna.cannahost",	XrmoptionSepArg,	NULL},
    {"-cannafile",	"*Canna.cannafile",	XrmoptionSepArg,	NULL},
#endif
#ifdef USE_SJ3
    {"-sj3",        ".conversionEngine",    XrmoptionNoArg,     "sj3"},
    {"-sj3serv",    "*Sj3.sj3serv",     XrmoptionSepArg,    NULL},
    {"-sj3serv2",   "*Sj3.sj3serv2",    XrmoptionSepArg,    NULL},
    {"-sj3user",    "*Sj3.sj3user",     XrmoptionSepArg,    NULL},
    {"-rcfile",     "*Sj3.rcfile",      XrmoptionSepArg,    NULL},
    {"-sbfile",     "*Sj3.sbfile",      XrmoptionSepArg,    NULL},
    {"-rkfile",     "*Sj3.rkfile",      XrmoptionSepArg,    NULL},
    {"-hkfile",     "*Sj3.hkfile",  XrmoptionSepArg,    NULL},
    {"-zhfile",     "*Sj3.zhfile",  XrmoptionSepArg,    NULL},
    {"-sjrc",       "*Sj3.rcfile",      XrmoptionSepArg,    NULL},
    {"-sjsb",       "*Sj3.sbfile",      XrmoptionSepArg,    NULL},
    {"-sjrk",       "*Sj3.rkfile",      XrmoptionSepArg,    NULL},
    {"-sjhk",       "*Sj3.hkfile",      XrmoptionSepArg,    NULL},
    {"-sjzh",       "*Sj3.zhfile",      XrmoptionSepArg,    NULL},
#endif
#ifdef USE_ATOK
    {"-atok",        ".conversionEngine",   XrmoptionNoArg,     "atok"},
    {"-atokserver",  "*Atok.server",    XrmoptionSepArg,    NULL},
    {"-as",          "*Atok.server",    XrmoptionSepArg,    NULL},
    {"-atokport",    "*Atok.port",      XrmoptionSepArg,    NULL},
    {"-atokconf",    "*Atok.conf",      XrmoptionSepArg,    NULL},
    {"-atokstyle",   "*Atok.style",     XrmoptionSepArg,    NULL},
#endif
};

XtAppContext	apc;
Widget		toplevel;

static int	numProtocols;
static int	(*DefaultErrorHandler)();

#if XtSpecificationRelease > 5
static XtSignalId	interrupt;
static void	interruptCallback();
#else
static void	exitTimer();
#endif

static WidgetClass	getInputObjClass();
static int	IgnoreBadWindow();
#ifdef SIGNALRETURNSINT
static int	scheduleExit();
#else
static void	scheduleExit();
#endif
static void	Destroyed();
static void	Exit();
static void	realExit();
static void	usage();
static void	print_version();

int
main(ac, av)
int ac;
char **av;
{
    Widget manager, protocol;
    int i;
    WidgetClass inputobjclass, displayobjclass;
    Widget inputobj;

    toplevel = XtAppInitialize(&apc, "Kinput2",
			       options, XtNumber(options),
			       &ac, av,
			       fallback_resources, (ArgList)NULL, 0);

    /* check invalid argument */ 
    if (ac > 1) {
	int do_usage = 0;
	for (i = 1; i < ac; i++) {
	    if (!strcmp(av[i], "-version")) {
		print_version();
	    } else {
		fprintf(stderr, "unknown argument: %s\n", av[i]);
		do_usage = 1;
	    }
	}
	if (do_usage) usage();
    }

    /* initialize asynchronous error handler */
    XAEInit();

    /* initialize I/O error callback handler */
    XIOEInit();

    XtGetApplicationResources(toplevel, &appres,
			      app_resources, XtNumber(app_resources),
			      NULL, 0);

    /*
     * If the application-specific class resource file
     * (the "app-defaults" file) is not found,
     * print a warning message.
     */
    if (!appres.appdefs_loaded) {
	fprintf(stderr, "Warning: Cannot load app-defaults file.\n");
	fprintf(stderr, "  Kinput2 may not work properly without it.\n");
	fprintf(stderr, "  Maybe kinput2 is not installed correctly,\n");
	fprintf(stderr, "  or your file search path (specified by\n");
	fprintf(stderr, "  environment variable 'XFILESEARCHPATH')\n");
	fprintf(stderr, "  is wrong.\n");
    }

    /* set debug level */
    debug_all = appres.debugLevel;

#ifdef RANDOM_ID
    /*
     * one nasty hack here:
     *
     * kinput clients often use server's window ID for the only key
     * value to identify their conversion server (kinput), and they
     * think it is dead and take appropriate action (eg connecting to
     * the new server) when they notice the ID has changed.
     *
     * but it is likely that another kinput has the same resource ID
     * base (because X servers always choose the smallest unused ID
     * base for new clients). and if it is the same, so is the owner
     * window ID, and the clients don't notice the change.
     *
     * to get rid of the problem, we add some small random offset to
     * the resource ID so that every time we get different owner ID
     * even if the resource ID base is the same.
     *
     * of course it heavily depends on the current implementaion of
     * the resource ID allocation in Xlib, so I call it 'nasty'.
     */
    XtDisplay(toplevel)->resource_id += getpid() % 1024;
#endif

    inputobjclass = getInputObjClass();

    inputobj = XtCreateWidget("inputObj", inputobjclass,
				toplevel, 0, 0);
    XtRealizeWidget(inputobj);
    ICRegisterTriggerKeys(inputobj);
    /*
       Destroying the `inputobj' is postponed until all the widgets
       are realized in order to prevent duplicate initialization of
       input object, that is, to prevent making connection twice to
       input conversion server.
     */

    displayobjclass = jpWcharDisplayObjectClass;

    manager = XtVaCreateManagedWidget("convmanager",
				      conversionManagerWidgetClass,
				      toplevel,
				      XtNwidth, 1,
				      XtNheight, 1,
				      NULL);

    numProtocols = 0;

    if (appres.useKinputProtocol) {
	protocol = XtVaCreateWidget("kinputprotocol",
				    kinputProtocolWidgetClass,
				    manager,
				    XtNlanguage, "JAPANESE",
				    XtNinputObjectClass, inputobjclass,
				    XtNdisplayObjectClass, displayobjclass,
				    XtNwidth, 1,
				    XtNheight, 1,
				    NULL);
	XtAddCallback(protocol, XtNdestroyCallback,
		      Destroyed, (XtPointer)NULL);
	numProtocols++;
    }

    if (appres.useXimpProtocol) {
	protocol = XtVaCreateWidget("ximpprotocol",
				    ximpProtocolWidgetClass,
				    manager,
				    XtNlocaleName, "ja_JP",
				    XtNinputObjectClass, inputobjclass,
				    XtNdisplayObjectClass, displayobjclass,
				    XtNwidth, 1,
				    XtNheight, 1,
				    NULL);
	XtAddCallback(protocol, XtNdestroyCallback,
		      Destroyed, (XtPointer)NULL);
	numProtocols++;
    }

    if (appres.useXIMProtocol) {
	protocol = XtVaCreateWidget("improtocol",
				    imProtocolWidgetClass,
				    manager,
				    XtNlanguage, "ja_JP",
				    XtNinputObjectClass, inputobjclass,
				    XtNdisplayObjectClass, displayobjclass,
				    XtNwidth, 1,
				    XtNheight, 1,
				    NULL);
	XtAddCallback(protocol, XtNdestroyCallback,
		      Destroyed, (XtPointer)NULL);
	numProtocols++;
    }

    if (numProtocols == 0) {
	fprintf(stderr, "no protocols activated\n");
	exit(1);
    }

    /* set signal handler */
    if (signal(SIGINT, SIG_IGN) != SIG_IGN) signal(SIGINT, scheduleExit);
#if XtSpecificationRelease > 5
    interrupt = XtAppAddSignal(apc, interruptCallback, (XtPointer)NULL);
#endif
    signal(SIGTERM, scheduleExit);
#ifdef USE_WNN
    signal(SIGPIPE, SIG_IGN);
#endif

    /* set my error handler */
    DefaultErrorHandler = XAESetErrorHandler(IgnoreBadWindow);

    XtRealizeWidget(toplevel);

    XtDestroyWidget(inputobj); /* Don't move this before XtRealizeWidget() */

    for (;;) {
	XEvent event;

	XtAppNextEvent(apc, &event);
	XtDispatchEvent(&event);
	MyDispatchEvent(&event); /* additional dispatcher */
    }
    /* NOTREACHED */

    return 0;	/* for lint */
}

static WidgetClass
getInputObjClass()
{
    WidgetClass class;

#ifdef USE_WNN
    if (!strcmp(appres.conversionEngine, "wnn")) {
	return ccWnnObjectClass;
    }
#endif
#ifdef USE_CANNA
    if (!strcmp(appres.conversionEngine, "canna") ||
	!strcmp(appres.conversionEngine, "iroha")) {
	return cannaObjectClass;
    }
#endif
#ifdef USE_SJ3
    if (!strcmp(appres.conversionEngine, "sj3")) {
	return sj3ObjectClass;
    }
#endif
#ifdef USE_ATOK
    if (!strcmp(appres.conversionEngine, "atok")) {
	return atokObjectClass;
    }
#endif

    /* set default input object */
#ifdef USE_ATOK
    class = atokObjectClass;
#endif
#ifdef USE_SJ3
    class = sj3ObjectClass;
#endif
#ifdef USE_CANNA
    class = cannaObjectClass;
#endif
#ifdef USE_WNN
    class = ccWnnObjectClass;
#endif

    return class;
}

static int
IgnoreBadWindow(dpy, error)
Display *dpy;
XErrorEvent *error;
{
    /*
     * BadWindow events will be sent if any of the active clients die
     * during conversion.  Although I select DestroyNotify event on the
     * client windows to detect their destruction and take appropriate
     * actions, and I'm careful when doing critical operations, but still
     * there is a chance of getting unexpecting BadWindow error caused by
     * client death.
     * There are also chances of getting BadDrawable as well.
     * So I set the error handler to ignore BadWindow/BadDrawable errors.
     * Of course I'd better check if the resourceid field of the error
     * event is the window ID of a client, but I'm too lazy to do that...
     * P.S.  We should ignore BadColor also..
     */
    if (error->error_code != BadWindow &&
	error->error_code != BadDrawable &&
	error->error_code != BadColor) {
	/* invoke default error handler */
	(*DefaultErrorHandler)(dpy, error);
    }
    return 0;
}

#ifdef SIGNALRETURNSINT
static int
#else
static void
#endif
scheduleExit()
{
#if XtSpecificationRelease > 5
    XtNoticeSignal(interrupt);
#else
    /*
     * It is unwise to do complex operation (in this case,
     * XtDestroyWidget) in a signal handler.
     * So postpone the real work...
     */
    XtAppAddTimeOut(apc, 1L, exitTimer, (XtPointer)NULL);
#endif
}

#if XtSpecificationRelease > 5
/* ARGSUSED */
static void
interruptCallback(cldata, sigid)
XtPointer cldata;
XtSignalId *sigid;
{
    Exit();
}
#else
/* ARGSUSED */
static void
exitTimer(cldata, timerp)
XtPointer cldata;
XtIntervalId *timerp;
{
    Exit();
}
#endif

/* ARGSUSED */
static void
Destroyed(w, client_data, call_data)
Widget w;
XtPointer client_data;
XtPointer call_data;
{
    static int cnt = 0;

    /*
     * if all the protocol widgets are dead, kill myself.
     */
    if (++cnt >= numProtocols) Exit();
}

static void
Exit()
{
    static int exiting = 0;

    TRACE(("Exit()\n"));

    if (exiting) return;
    exiting = 1;

    /*
     * Destroy all widgets.
     */
    XtDestroyWidget(toplevel);

    /*
     * Postpone calling exit() until next XtNextEvent(),
     * in order to give each widget time to execute their
     * destroy procedure.
     */
    XtAppAddTimeOut(apc, 1L, realExit, (XtPointer)NULL);
}

/* ARGSUSED */
static void
realExit(cldata, timerp)
XtPointer cldata;
XtIntervalId *timerp;
{
    exit(0);
}

static void
usage()
{
    char **p;
    static char *syntaxtable[] = {
#ifdef USE_WNN
	"-wnn",			"use Wnn as the conversion engine",
	"-jserver <hostname>",	"specify jserver host",
	"-ccdef <ccdeffile>",	"specify character conversion def. file",
	"-wnnenvname <name>",	"specify Wnn environment name",
	"-wnnenvrc4 <wnnenvrcfile>", "specify Wnn environment file for Wnn4",
	"-wnnenvrc6 <wnnenvrcfile>", "specify Wnn environment file for Wnn6",
	"-wnnenvrc <wnnenvrcfile>", "specify Wnn environment file",
#endif
#ifdef USE_CANNA
	"-canna",		"use Canna (Iroha) as the conversion engine",
	"{-cannaserver|-cs} <hostname>[:n]", "specify cannaserver host",
	"-cannafile <cannafile>", "specify canna customize file",
#endif
#ifdef USE_SJ3
	"-sj3",         	"use SJ3 as the conversion engine",
	"-sj3serv <hostname>",  "specify first sj3serv host",
	"-sj3serv2 <hostname>", "specify second sj3serv host",
	"-sj3user <user>",      "specify user name connect to sj3serv",
	"{-rcfile|-sjrc} <file>",       "specify resource definition file",
	"{-sbfile|-sjsb} <file>",       "specify symbol table file",
	"{-rkfile|-sjrk} <file>",       "specify roma-kana coversion definition file",
	"{-hkfile|-sjhk} <file>",       "specify hira-kata coversion definition file",
	"{-zhfile|-sjzh} <file>",       "specify zen/han coversion definition file",
#endif
#ifdef USE_ATOK
	"-atok",                        "use ATOK as the conversion engine",
	"{-atokserver|-as} <hostname>", "specify atok server host",
	"-atokport <port#>",            "specify atok service port",
	"-atokconf <file>",             "specify atok customize file",
	"-atokstyle <file>",            "specify atok style file",
#endif
	"-bc",			"backward compatible mode",
	"-font <font>",		"ASCII font to be used",
	"-kanjifont <font>",	"KANJI font to be used",
	"-kanafont <font>",	"KANA font to be used",
	"-/+kinput",		"activate/deactivate kinput protocol family",
	"-/+ximp",		"activate/deactivate Ximp protocol",
	"-/+xim",		"activate/deactivate X Input Method protocol",
	"-tbheight <number>",	"specify pop-up shell's titlebar height",
	"-background <color>",	"background color",
	"-foreground <color>",	"foreground color",
	"-rv",			"reverse video mode",
	"-display <display>",	"specify display",
	"-version",		"print version information and exit",
#ifdef DEBUG
	"-debug",		"print debug messages (debug level 1)",
	"-trace",		"print trace messages (debug level 10)",
	"-debuglevel <level>",	"set debug level",
#endif
	NULL, NULL,
    };

    fprintf(stderr, "options are:\n");
    for (p = syntaxtable; *p != NULL; p += 2) {
	fprintf(stderr, "    %-30s %s\n", *p, *(p + 1));
    }
    exit(1);
}

static void
print_version()
{
    char *p;

    printf("kinput2 %s ", KINPUT2_VERSION);
    if (PATCHLEVEL > 0) printf("fix %d ", PATCHLEVEL);
#ifdef STATUS
    printf("-%s- ", STATUS);
#endif
    printf(" (");
    p = DATE + 7;				/* skip '$Date: ' */
    while (*p != '\0' && *p != ' ') {
	putchar(*p);	/* print date */
	p++;
    }
    printf(")\n");

    printf("\toptions: ");
#ifdef USE_WNN
#ifdef USE_WNN6
    printf("[Wnn6] ");
#else
    printf("[Wnn] ");
#endif
#endif
#ifdef USE_CANNA
    printf("[Canna2] ");
#endif
#ifdef USE_SJ3
    printf("[Sj3] ");
#endif
#ifdef USE_ATOK
    printf("[Atok] ");
#endif
#ifdef DEBUG
    printf("[DEBUG] ");
#endif
    printf("\n");
    exit(0);
}

#if defined(USE_WNN) && defined(NEED_Strlen)
/*
 * Wnn/jlib/js.c should have this function...
 */
int
Strlen(s)
unsigned short *s;
{
    int n = 0;

    while (*s++) n++;
    return n;
}
#endif
